### PR TITLE
feat: Instance Sizes Feature

### PR DESCRIPTION
**Pull request by wtrocki**
Added initial PR with the all hooks that we require to get things working.

Still early version. I need to refactor it little bit to use underlying types etc.

I need feedback on:

1. Do we like this architecture of single hook per file vs api.ts (hate api.ts as collection of hooks)
2. How we can share kafka instance type across hooks. I think AMS hook is too low level we need to get type once and then reuse it across hooks but that affects some component logic :/
3. This works end to end but it has dozens of TODOs
4. Can we remove CreateInstance (There is larger refactoring/cleanup needed to remove trial and keeping old version prevent us from doing that.

### CODE REVIEW

**Comment by wtrocki** [comment]
I have added quota hook into kas-ui to resolve some issues we have been seeing and also for clarity. This is still not fully functional. I need to resolve some TODOs

File: src/app/modules/OpenshiftStreams/dialogs/CreateInstance/CreateInstanceWithSizes/hooks/useAMSQuota.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+// Helper methods for busines logic handled in the UI
+export function convertQuotaToInstanceType(quota: Map<QuotaType, QuotaValue> | undefined) {
+  let kasQuota: QuotaValue | undefined;
+  try {
+    kasQuota = quota?.get(QuotaType?.kas);
+  } catch (e) {
+    console.error('useAvailableProvidersAndDefault', 'quota?.get exception', e);
+  }
+  if (kasQuota !== undefined && kasQuota.remaining > 0) {
```
Comment (ajaypratap003) [inline src/app/modules/OpenshiftStreams/dialogs/CreateInstance/CreateInstanceWithSizes/hooks/useAMSQuota.ts:88]:
```suggestion
  if (kasQuota !== undefined && kasQuota.remaining >= 0) {
```
I think InstanceType should be standard if user has quota `kasQuota.remaining`. @wtrocki correct me if I'm wrong.
`kasQuota.remaining` count could be 0 if user consumed all the limit then still InstanceType  should be standard. 

InstanceType `developer` should be only when user don't have standard quota .

File: src/app/modules/OpenshiftStreams/dialogs/CreateInstance/CreateInstanceWithSizes/hooks/useAMSQuota.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+export function getQuotaType(quota: Map<QuotaType, QuotaValue> | undefined) {
+  let kasQuota: QuotaValue | undefined;
+  try {
+    kasQuota = quota?.get(QuotaType?.kas);
+  } catch (e) {
+    console.error('useAvailableProvidersAndDefault', 'quota?.get exception', e);
+  }
+
+  if (kasQuota !== undefined && kasQuota.remaining > 0) {
```
Comment (ajaypratap003) [inline src/app/modules/OpenshiftStreams/dialogs/CreateInstance/CreateInstanceWithSizes/hooks/useAMSQuota.ts:103]:
```suggestion
  if (kasQuota !== undefined && kasQuota.remaining >= 0) {
```

File: src/app/modules/OpenshiftStreams/dialogs/CreateInstance/CreateInstanceWithSizes/hooks/useAvailableProvidersAndDefault.ts [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+            };
+          })
+      );
+    } catch (e) {
+      console.error('useAvailableProvidersAndDefault', 'fetchProvider', e);
+      return Promise.reject(e);
+    }
+  };
+
+  // TODO - this logic should not exist and be replaced with AMS quota.
```
Comment (ajaypratap003) [inline src/app/modules/OpenshiftStreams/dialogs/CreateInstance/CreateInstanceWithSizes/hooks/useAvailableProvidersAndDefault.ts:84]: This logic used for trial kafka to restrict user to create single instance on create modal.
We can remove this logic and use AMS quota if  AMS quota set the limit for trial/developer. 
AFAIK, AMS quota doesn't set the limit for trail, they always have zero limit. 
Do know about AMS quota limit for trial/developer ?

File: src/app/modules/OpenshiftStreams/dialogs/CreateInstance/CreateInstanceWithSizes/hooks/useAvailableProvidersAndDefault.ts [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+      console.error('useAvailableProvidersAndDefault', 'fetchUserHasTrialInstance', e);
+    }
+    return false;
+  };
+
+  return async function (): Promise<CreateKafkaInstanceWithSizesTypes.CreateKafkaInitializationData> {
+    try {
+      const quota = await getQuota();
+      const plan = getQuotaType(quota.data);
+      const instance_type = convertQuotaToInstanceType(quota.data);
```
Comment (ajaypratap003) [inline src/app/modules/OpenshiftStreams/dialogs/CreateInstance/CreateInstanceWithSizes/hooks/useAvailableProvidersAndDefault.ts:107]:
```suggestion
      const instanceType = convertQuotaToInstanceType(quota.data);
```

File: src/app/modules/OpenshiftStreams/dialogs/CreateInstance/CreateInstanceWithSizes/hooks/useAMSQuota.ts (hunk 1 above)
Comment (wtrocki) [reply src/app/modules/OpenshiftStreams/dialogs/CreateInstance/CreateInstanceWithSizes/hooks/useAMSQuota.ts:88]: I actually copied that code from other place. Not sure if remaining should be used. 
Curent BU requirement is to not allow trial instances to be created even if user runs out of quota. 

kasQuota.remaining >= 0 will be always true as there is no possibility of negative values. I think we might need to remove it.

Comment (wtrocki) [reply src/app/modules/OpenshiftStreams/dialogs/CreateInstance/CreateInstanceWithSizes/hooks/useAMSQuota.ts:88]: This logic reflects very well bu requirement: https://github.com/bf2fc6cc711aee1a0c2a/kas-ui/blob/4842ab5d11df30a91af77d6e275e0adabc7473db/src/app/modules/OpenshiftStreams/dialogs/CreateInstance/api.ts#L207-L209

File: src/app/modules/OpenshiftStreams/dialogs/CreateInstance/CreateInstanceWithSizes/hooks/useAvailableProvidersAndDefault.ts (hunk 3 above)
Comment (wtrocki) [reply src/app/modules/OpenshiftStreams/dialogs/CreateInstance/CreateInstanceWithSizes/hooks/useAvailableProvidersAndDefault.ts:84]: We cannot avoid having different logic for trial and standard instances sadly or use any AMS values for trial.

Comment (wtrocki) [reply src/app/modules/OpenshiftStreams/dialogs/CreateInstance/CreateInstanceWithSizes/hooks/useAvailableProvidersAndDefault.ts:84]: > We can remove this logic and use AMS quota if AMS quota set the limit for trial/developer.

Would explore that idea

File: src/app/modules/OpenshiftStreams/dialogs/CreateInstance/CreateInstanceWithSizes/hooks/useAvailableProvidersAndDefault.ts (hunk 4 above)
Comment (wtrocki) [reply src/app/modules/OpenshiftStreams/dialogs/CreateInstance/CreateInstanceWithSizes/hooks/useAvailableProvidersAndDefault.ts:107]: ignore some nit/names for now. This code is copied over from other places. I did not write any of this, just reorganized. Will improve it once we have logic in place.

**Comment by riccardo-forina** [comment]
This is not fully done yet, it requires checking for the various error codes from the APIs to match them to the right UI state.
We'll do it in a follow-up PR.

**Pull request merged**